File: app/src/main/java/de/westnordost/streetcomplete/quests/way_lit/AddWayLit.java
package de.westnordost.streetcomplete.quests.way_lit;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import java.util.Arrays;
import java.util.Map;

import javax.inject.Inject;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao;
import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment;
import de.westnordost.streetcomplete.quests.YesNoQuestAnswerFragment;

public class AddWayLit extends SimpleOverpassQuestType
{
	private static final String[] LIT_RESIDENTIAL_ROADS = { "residential", "living_street", "pedestrian" };

	private static final String[] LIT_NON_RESIDENTIAL_ROADS = {
			"primary", "secondary", "tertiary", "unclassified", "service",
	};

	private static final String[] LIT_WAYS = { "footway", "cycleway", "steps" };

	@Inject public AddWayLit(OverpassMapDataDao overpassServer) { super(overpassServer); }

	@Override protected String getTagFilters()
	{
		/* Using sidewalk as a tell-tale tag for (urban) streets which reached a certain level of
		   development. I.e. non-urban streets will usually not even be lit in industrialized
		   countries.
		   Also, only include paths only for those which are equal to footway/cycleway to exclude
		   most hike paths and trails.

		   See #427 for discussion. */

		return "ways with " +
				"(" +
				"  highway ~ " + TextUtils.join("|", LIT_RESIDENTIAL_ROADS) +
				"  or" +
				"  highway ~ " + TextUtils.join("|", LIT_NON_RESIDENTIAL_ROADS) + " and" +
				"  (" +
				"    sidewalk ~ both|left|right|yes|separate" +
				"    or source:maxspeed ~ .+:urban or maxspeed:type ~ .+:urban or zone:maxspeed ~ .+:urban" +
				"  )" +
				"  or" +
				"  highway ~ " + TextUtils.join("|", LIT_WAYS) +
				"  or" +
				"  highway = path and (foot = designated or bicycle = designated)" +
				")" +
				" and !lit" +
				" and (access !~ private|no or (foot and foot !~ private|no))"; // not private roads
	}

	public AbstractQuestAnswerFragment createForm()
	{
		return new WayLitForm();
	}

	public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes)
	{
		String other = answer.getString(WayLitForm.OTHER_ANSWER);
		if (other != null)
		{
			changes.add("lit", other);
		}
		else
		{
			changes.add("lit", answer.getBoolean(YesNoQuestAnswerFragment.ANSWER) ? "yes" : "no");
		}
	}

	@Override public String getCommitMessage() { return "Add whether way is lit"; }
	@Override public int getIcon() { return R.drawable.ic_quest_lantern; }
	@Override public int getTitle(@NonNull Map<String,String> tags)
	{
		String type = tags.get("highway");
		boolean hasName = tags.containsKey("name");
		boolean isRoad = Arrays.asList(LIT_NON_RESIDENTIAL_ROADS).contains(type) ||
				Arrays.asList(LIT_RESIDENTIAL_ROADS).contains(type);

		if (hasName)      return R.string.quest_way_lit_named_title;
		else if (isRoad)  return R.string.quest_way_lit_road_title;
		else              return R.string.quest_way_lit_title;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/way_lit/AddWayLit.kt
package de.westnordost.streetcomplete.quests.way_lit

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osm.geometry.ElementGeometry
import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement.PEDESTRIAN
import de.westnordost.streetcomplete.osm.MAXSPEED_TYPE_KEYS
import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
import de.westnordost.streetcomplete.data.quest.DayNightCycle.ONLY_NIGHT
import de.westnordost.streetcomplete.osm.Tags
import de.westnordost.streetcomplete.osm.changeToSteps
import de.westnordost.streetcomplete.osm.lit.applyTo

class AddWayLit : OsmFilterQuestType<WayLitOrIsStepsAnswer>() {

    /* Using sidewalk, source:maxspeed=*urban etc and a urban-like maxspeed as tell-tale tags for
       (urban) streets which reached a certain level of development. I.e. non-urban streets will
       usually not even be lit in industrialized countries.

       Also, only include paths only for those which are equal to footway/cycleway to exclude
       most hike paths and trails.

        See #427 for discussion. */
    override val elementFilter = """
        ways with
        (
          highway ~ ${LIT_RESIDENTIAL_ROADS.joinToString("|")}
          or highway ~ ${LIT_NON_RESIDENTIAL_ROADS.joinToString("|")} and
          (
            sidewalk ~ both|left|right|yes|separate
            or sidewalk:both = yes
            or sidewalk:left = yes
            or sidewalk:right = yes
            or ~"${(MAXSPEED_TYPE_KEYS + "maxspeed").joinToString("|")}" ~ ".*:(urban|.*zone.*|nsl_restricted)"
            or maxspeed <= 60
          )
          or highway ~ ${LIT_WAYS.joinToString("|")}
          or highway = path and (foot = designated or bicycle = designated)
        )
        and (access !~ private|no or (foot and foot !~ private|no))
        and
        (
          !lit
          or lit = no and lit older today -8 years
          or lit older today -16 years
        )
        and indoor != yes
        and ~path|footway|cycleway !~ link
    """

    override val changesetComment = "Specify whether ways are lit"
    override val wikiLink = "Key:lit"
    override val icon = R.drawable.ic_quest_lantern
    override val achievements = listOf(PEDESTRIAN)
    override val defaultDisabledMessage = R.string.default_disabled_msg_overlay
    override val dayNightCycle = ONLY_NIGHT

    override fun getTitle(tags: Map<String, String>) = R.string.quest_lit_title

    override fun createForm() = WayLitForm()

    override fun applyAnswerTo(answer: WayLitOrIsStepsAnswer, tags: Tags, geometry: ElementGeometry, timestampEdited: Long) {
        when (answer) {
            is IsActuallyStepsAnswer -> tags.changeToSteps()
            is WayLit -> answer.litStatus.applyTo(tags)
        }
    }

    companion object {
        private val LIT_RESIDENTIAL_ROADS = arrayOf("residential", "living_street", "pedestrian")

        private val LIT_NON_RESIDENTIAL_ROADS = arrayOf(
            "motorway", "motorway_link", "trunk", "trunk_link", "primary", "primary_link",
            "secondary", "secondary_link", "tertiary", "tertiary_link", "unclassified", "service"
        )

        private val LIT_WAYS = arrayOf("footway", "cycleway", "steps")
    }
}
